### ISSUE TITLE
Uncalled function checks don't work when the variable is referenced in the block

### ISSUE THREAD

**Issue by jonhue** [microsoft/TypeScript#43211]
# Bug Report

### 🔎 Search Terms

`uncalled function checks`, `conjunction`, `&&`

### 🕗 Version & Regression Information

This is the behavior in every version I tried

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=21&pc=2#code/PTAEAEGcBcCcEsDG0ByBXANhgwgCwKaIDWkAXKHGvgFDUAmhGAhrPqAGZoB2y8A9l1DxIAMT58AFAEpyAI3EZ8TLgG56jFm049o-QcIBCLaXIVLVteO1AThYvlNABvaqDegQoboiZZ8dDm9dAVBEAmJIUF9WJjoAT1AAdz5YIiT8LFACVmoAX0trW1FxUAAyUqFII1hpRxd3DzBvX0UA7V4QsMISKIwY+KSUtMSMjCz8HPzqKxs7EvLK+1rnV3dPZr82oL1Q8J7IXD5MAMQ+ADcJilxhUKZINmhxPILZ4r4yisNjKTrVtzm+GoGuseC1-IEdDsuhFQAcjhgTudLtBrpEfPcKE9ckA)

### 💻 Code

```ts
// @strictNullChecks: true

declare function isFoo(): boolean;
declare function isBar(): boolean;

if (isFoo) {
    // uncalled function checks already work well here
}

if (isFoo && isBar()) {
    // uncalled function checks already work well here
}

if (isFoo && isFoo()) {
    // uncalled function checks should cover this case too
}

if (isFoo && isBar()) {
    isFoo;
    // uncalled function checks should cover this case too
}
```

### 🙁 Actual behavior

In the last two cases we didn't detect that the first occurrence of `isFoo` in the condition is uncalled.

### 🙂 Expected behavior

When `isFoo` cannot possibly be a falsy value (i.e. `undefined`) we should highlight this as an error even if `isFoo` occurs again later on. This is a follow-up to a conversation in https://github.com/microsoft/TypeScript/pull/42835#discussion_r590868388.
I would be happy to enable this check, then we could run this against some user tests and decide whether there are too many false positives 🙂

**Comment by RyanCavanaugh** [comment]
A reference should be sufficient here to not complain, since you might write something like
```ts
if (isFoo) {
  invokeThisPlease(isFoo);
}
```

**Comment by jonhue** [comment]
@RyanCavanaugh I'm probably misunderstanding your example, but if `isFoo` cannot possible be `undefined`, isn't the condition in the example you gave above redundant?

**Comment by RyanCavanaugh** [comment]
It's intentional that the rule is not ironclad. This is because there are some places in the language where TS thinks something is always not-`undefined`, but it's not necessarily the case.
